app/javascript/utils/highlight.ts

```diff
@@ -0,0 +1,71 @@
+import * as highlighter from 'highlight.js'
+
+function duplicateMultilineNodes(element: HTMLElement) {
+  element.childNodes.forEach((child) => {
+    if (child.textContent === null) {
+      return
+    }
+
+    if (getLinesCount(child.textContent) === 0) {
+      return
+    }
+
+    if (child.childNodes.length === 0) {
+      return duplicateMultilineNode(child.parentNode as HTMLElement)
+    }
+
+    duplicateMultilineNodes(child as HTMLElement)
+  })
+}
+
+function duplicateMultilineNode(element: HTMLElement) {
+  const isHighlightJSNode = /hljs-/.test(element.className)
+
+  if (!isHighlightJSNode) {
+    return
+  }
+
+  element.innerHTML = getLines(element.innerHTML)
+    .map(
+      (line) =>
+        `<span class="${element.className}">${
+          line.length > 0 ? line : ' '
+        }</span>\n`
+    )
+    .join('')
+    .trim()
+
+  return
+}
+
+function getLines(text: string) {
+  return text.split(/\r\n|\r|\n/g)
+}
+
+function getLinesCount(text: string) {
+  return (text.trim().match(/\r\n|\r|\n/g) || []).length
+}
+
+function wrapLineNumbers(code: string) {
+  const element = document.createElement('div')
+  element.innerHTML = code.trim()
+
+  duplicateMultilineNodes(element)
```

Comment:
Highlight JS wraps multiline nodes within one parent element. This wouldn't work in our case since we need to append line numbers to each. This method recursively expands the tree and adds wraps each line in the multiline node with the correct class.

Reply:
This should be a comment in the code.